File: problems/tier2-intermediate/020-logger-system/boilerplate/java/part2/Guided.java
import java.util.*;

// Data class (given).
class LogOp {
    public String kind;
    public String s1;
    public String s2;
    public int i1;

    public LogOp(String kind, String s1, String s2, int i1) {
        this.kind = kind;
        this.s1 = s1;
        this.s2 = s2;
        this.i1 = i1;
    }

    public LogOp(String kind) {
        this(kind, "", "", 0);
    }
}

// HINT: introduce an abstraction so new variants don't change existing code.
public class Solution {
    // HINT: start from what this must return, then work backwards to the state it needs.
    public static List<String> logger_simulate(List<LogOp> ops) {
        // TODO: write your solution
        return null;
    }

}

File: problems/tier2-intermediate/020-logger-system/boilerplate/cpp/part1/guided.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;


// Data class (given).
struct LogOp {
    string kind;
    string s1;
    string s2;
    int i1;
    LogOp(const string& kind_, const string& s1_ = "", const string& s2_ = "", int i1_ = 0)
      : kind(kind_), s1(s1_), s2(s2_), i1(i1_) {}
};

// HINT: introduce an abstraction so new variants don't change existing code.
// HINT: keep each piece small — one responsibility per class.

// HINT: start from what this must return, then work backwards to the state it needs.
vector<string> logger_simulate(vector<LogOp> ops) {
    // TODO: write your solution
    return {};
}
